src/basic/fs-util.h

```diff
@@ -14,6 +14,13 @@
 #include "errno-util.h"
 #include "time-util.h"
 
+#define MODE_INVALID ((mode_t) -1)
+
+/* The following macros add 1 when converting things, since 0 is a valid mode, while the pointer
+ * NULL is special */
+#define PTR_TO_MODE(p) ((mode_t) ((uintptr_t) (p)-1))
+#define MODE_TO_PTR(u) ((void *) ((uintptr_t) (u)+1))
```

Comment:
Shouldn't MODE_INVALID be handled specially, i.e. not have +1 added?

Reply:
No, here +1 or -1 intends NULL <-> MODE_INVALID, that is, non-NULL means valid mode.
We already do the same thing in `UID_TO_PTR()`/`PTR_TO_UID()`.